### ISSUE TITLE
ValueError: unsupported pickle protocol: 5 while ensemble=True

### ISSUE THREAD

**Issue by prateekgml** [microsoft/FLAML#680]
Hi,

I am trying to use ensemble method while training a dataset as below-

![image](https://user-images.githubusercontent.com/97145738/184355790-ecd0fc83-2cde-49f0-b2f9-46bcb623f09d.png)

The training was fine but in the end of the following line- `[flaml.automl: 08-12 12:34:01] {3407} WARNING - Using passthrough=False for ensemble because the data contain categorical features.`

I am getting below -

![image](https://user-images.githubusercontent.com/97145738/184356015-a7fe3306-f736-47d5-827e-01b8b1d3b99b.png)

And then the following error occurred-

![image](https://user-images.githubusercontent.com/97145738/184356687-914502eb-2c8e-4380-bc34-c54dbca31a3e.png)

![image](https://user-images.githubusercontent.com/97145738/184356110-52f5f468-2d53-4e7d-bfea-c1943e18dc86.png)

Without ensemble, I was able to complete the training.

Please help me to resolve this issue.

**Comment by sonichi** [comment]
The error message is https://user-images.githubusercontent.com/97145738/184356110-52f5f468-2d53-4e7d-bfea-c1943e18dc86.png
It suggests that you don't have enough RAM to build the ensemble. You can try specifying a simple final_estimator, e.g.,
```python
automl.fit(
    X_train, y_train, task="classification",
    "ensemble": {
        "final_estimator": LogisticRegression(),
        "passthrough": False,
    },
)
```

**Comment by prateekgml** [comment]
@sonichi is this syntax is correct-

![image](https://user-images.githubusercontent.com/97145738/184467286-57dde562-80d7-46c1-9e56-f9911542763b.png)

And what should I import to resolve this error with above syntax-

![image](https://user-images.githubusercontent.com/97145738/184467303-a6e69fcd-d2c8-40a8-ba30-fd0f3b3a8b8b.png)

**Comment by prateekgml** [comment]
I am able to proceed after importing as `from sklearn.linear_model import LogisticRegression`

**Comment by prateekgml** [comment]
Now training is completed but RAM error is occurred as below-

![image](https://user-images.githubusercontent.com/97145738/184477158-e95e6ac8-cd60-4c21-92da-6f52ad648e6e.png)

It seems ensembling is not possible with FLAML.

**Comment by sonichi** [comment]
> Now training is completed but RAM error is occurred as below-
> 
> ![image](https://user-images.githubusercontent.com/97145738/184477158-e95e6ac8-cd60-4c21-92da-6f52ad648e6e.png)
> 
> It seems ensembling is not possible with FLAML.

How large is the dataset and how large is the free RAM?
Ensemble requires more RAM than training each single model. One thing you can try is to use a smaller time budget for tuning such that small models will be used for ensembling.

**Comment by prateekgml** [comment]
> > Now training is completed but RAM error is occurred as below-
> > ![image](https://user-images.githubusercontent.com/97145738/184477158-e95e6ac8-cd60-4c21-92da-6f52ad648e6e.png)
> > It seems ensembling is not possible with FLAML.
> 
> How large is the dataset and how large is the free RAM? Ensemble requires more RAM than training each single model. One thing you can try is to use a smaller time budget for tuning such that small models will be used for ensembling.

Dataset is not so big as you can see here-

![image](https://user-images.githubusercontent.com/97145738/184625353-72d55a75-979e-4ecd-a88f-8d11d8a7e5e5.png)

With `"time_budget\": 10000*10,` I am able to continue the training in Kaggle. Since its execution time crossed 12 hours limit, it stopped further.

I will continue my experiment with Colab Pro+ which has Tesla P100 16 GB GPU and 54 GiB RAM. I will inform you once the investigation is done.

**Comment by sonichi** [comment]
> > > Now training is completed but RAM error is occurred as below-
> > > ![image](https://user-images.githubusercontent.com/97145738/184477158-e95e6ac8-cd60-4c21-92da-6f52ad648e6e.png)
> > > It seems ensembling is not possible with FLAML.
> > 
> > 
> > How large is the dataset and how large is the free RAM? Ensemble requires more RAM than training each single model. One thing you can try is to use a smaller time budget for tuning such that small models will be used for ensembling.
> 
> Dataset is not so big as you can see here-
> 
> ![image](https://user-images.githubusercontent.com/97145738/184625353-72d55a75-979e-4ecd-a88f-8d11d8a7e5e5.png)
> 
> With `"time_budget\": 10000*10,` I am able to continue the training in Kaggle. Since its execution time crossed 12 hours limit, it stopped further.
> 
> I will continue my experiment with Colab Pro+ which has Tesla P100 16 GB GPU and 54 GiB RAM. I will inform you once the investigation is done.

Thanks. Have you tried using smaller time budget to test whether the ensemble works?
One other question is did you feed the raw data into `AutoML` or did you do one-hot encoding before it? One-hot encoding could blow up the size of the dataset and should be avoided.

**Comment by prateekgml** [comment]
Yes, I started the experiment with `3600` but then ensembling was covering 96% so I increased the time budget.
And I am feeding raw data into the FLAML.

**Comment by sonichi** [comment]
> I mean with a lower time budget, search was 96% and in log message it was saying to increase the time budget to complete the search as 100%.

Did ensemble succeed with lower time budget?

**Comment by prateekgml** [comment]
No, it was not successful.

**Comment by prateekgml** [comment]
Please suggest some steps to improve the result using ensembling. My current settings are as below-

![image](https://user-images.githubusercontent.com/97145738/184803946-34501081-d285-4f1d-a0e4-309985fc0945.png)

**Comment by sonichi** [comment]
If you have enough RAM now, try removing the key "final_estimator" from "ensemble".

**Comment by prateekgml** [comment]
I have started the training with below changes in colab pro+ which have 54 GiB RAM-

![image](https://user-images.githubusercontent.com/97145738/184865371-9e926d79-51ce-4b71-94c1-8bef06218364.png)

**Comment by sonichi** [comment]
Use `log_type="all"`. https://microsoft.github.io/FLAML/docs/Use-Cases/Task-Oriented-AutoML#log-the-trials

**Comment by prateekgml** [comment]
I did an experiment with commenting on the ensemble line like below-

![image](https://user-images.githubusercontent.com/97145738/184873950-568df99f-a31b-4754-af32-d51e8bc4054a.png)

And the score dropped to 83.3.

**Comment by prateekgml** [comment]
@sonichi I tried adding more estimators with the ensemble as below-

![image](https://user-images.githubusercontent.com/97145738/185102196-cf6e61ac-9eef-48a8-9b1e-8dfbe18a1cb8.png)

And my score improved to 83.968 from 83.55. Is there anything I can try to improve the score?

**Comment by sonichi** [comment]
> @sonichi I tried adding more estimators with the ensemble as below-
> 
> ![image](https://user-images.githubusercontent.com/97145738/185102196-cf6e61ac-9eef-48a8-9b1e-8dfbe18a1cb8.png)
> 
> And my score improved to 83.968 from 83.55. Is there anything I can try to improve the score?

What about using all the estimators, by removing the `estimator_list` argument?

**Comment by prateekgml** [comment]
After removing the estimator_list, the score dropped to 83.92 from 83.96.
